### ISSUE TITLE
[board-server] Allow Board Server to run in a container as a self-contained deployment 

### ISSUE THREAD

**Issue by wfaithfull** [breadboard-ai/breadboard#2889]
The Board Server is an incredibly exciting development for the share-ability of boards, but currently the means of deploying it is very GCP-centric, and the storage depends on Firestore. I would like to:

- localise the storage out of firestore - maybe into SQLite or similar.
- write a dockerfile that builds a completely self-contained board server container.

**Comment by timswanson-google** [comment]
@aomarks has been looking at this

**Comment by wfaithfull** [comment]
Is there a duplicate issue? I didn't spot one. @aomarks let me know what you've already done towards this - I've got a branch going but I will happily defer that and/or contribute into work you already have if you are further along this objective.

**Comment by wfaithfull** [comment]
Bulk of my changes are [here](https://github.com/ExaDev/breadboard/blob/2889-board-server-sqlite/packages/board-server/src/server/store.ts#L342): reimplementing `Store` with a SQLite backend instead of firestore. ⚠️ here be dragons, very much WIP ⚠️ 

I've focussed exclusively so far on breaking the dependency on firestore than on containerising - because I want the whole thing to be able to be self-contained as a deployment. For the container question I think it's a question of where we intend to publish the resulting image from - if we publish it from a github action here into GHCR.io, I think it definitely makes sense that we take the github approach. If I was going to publish a dockerfile as an independent third party I might differ.

**Comment by timswanson-google** [comment]
Tangential to this issue (specifically wrt to not depending on Firestore), I've filed #2896 to stop using Firestore for CORS config.

**Comment by wfaithfull** [comment]
First successful test run of a fully self-contained, containerised board server with SQLite storage, running in a container. 
```
(base) ➜  board-server git:(2889-board-server-sqlite) ✗ docker run -e GOOGLE_CLOUD_PROJECT=foo -p 3000:3000 board-server

> @breadboard-ai/board-server@0.4.6 serve
> wireit --host=0.0.0.0

Analyzing
 71% [12 / 17] [0 running] [0 services]
Running on "http://localhost:3000"...
100% [17 / 17] [1 running] [1 service] serve
```
```
(base) ➜  board-server git:(2889-board-server-sqlite) ✗ curl -s localhost:3000/boards | jq 
[
  {
    "title": "test",
    "path": "wfaithfull/test.json",
    "username": "wfaithfull",
    "readonly": true,
    "mine": false,
    "tags": "published"
  }
]
```

For the moment, my dockerfile definition looks like so:

```
FROM node:20
COPY --from=breadboard . /breadboard/
WORKDIR /breadboard
RUN npm ci
WORKDIR /breadboard/packages/board-server
RUN npm run build
CMD ["npm", "run", "serve", "--", "--host=0.0.0.0"]
```

This build context malarky is a little ugly but preferable in order to allow the dockerfile to build within the context of the board-server package (since docker doesn't allow straightforward operations like `../` in the build context). You can build it like so:

```
docker build --build-context breadboard=../../ -t board-server .
```

I'll refine this a bit more but it's a good start.

**Comment by wfaithfull** [comment]
Some updates since I just merged in @timswanson-google's changes adding a `ServerConfig` object. 

So I'm adding some integration tests, for that reason I've broken out the "server start" machinery into a separate file so that I can call it from tests to bootstrap a server to test against:

https://github.com/ExaDev/breadboard/blob/2889-board-server-sqlite/packages/board-server/src/index.ts

Secondly, I've added some global error handling & 500 responses to the router so that the whole thing doesn't crash on unhandled exceptions.

https://github.com/ExaDev/breadboard/blob/2889-board-server-sqlite/packages/board-server/src/router.ts#L16

**Comment by timswanson-google** [comment]
Thanks Will. Can we get that minimal change ("server start" to server.ts) pulled into the breadboard-ai main branch? I'd like to make a change on top of it for #2896.

**Comment by wfaithfull** [comment]
I have to go to something in 15 minutes I'm afraid so I'm not going to get around to that until tomorrow... you can pick it off my branch if you like and roll it in from your branch? I am happy to accept the merge pain 😄

**Comment by timswanson-google** [comment]
Will do, thanks

**Comment by timswanson-google** [comment]
Merged #2905 into main. Since I'm no git wizard, I did it the old fashioned way using copy-paste.
